File: lista4/go/main.go
package main

import (
	"sync"
	"go/philosopher"
	"go/fork"
)

const numberOfPhilosophers = 3
const mealLimit = 3

var waitGroup = sync.WaitGroup{}

func main() {
	var forks [numberOfPhilosophers]fork.Fork
	var philosophers [numberOfPhilosophers]*philosopher.Philosopher

	for i := 0; i < numberOfPhilosophers; i++ {
		lowerForkIndex := i
		higherForkIndex := (i + 1) % numberOfPhilosophers

		if lowerForkIndex > higherForkIndex {
			lowerForkIndex, higherForkIndex = higherForkIndex, lowerForkIndex
		}

		philosophers[i] = &philosopher.Philosopher{
			Id:			i + 1,
			LowerFork:	&forks[lowerForkIndex],
			HigherFork: &forks[higherForkIndex],
			MealsCount: mealLimit,
			WaitGroup:			&waitGroup,
		}
	}

	for _, philosopher := range philosophers {
		waitGroup.Add(1)
		go philosopher.Dine()
	}

	waitGroup.Wait()
}

File: lista4/python/main.py
from fork import Fork
from philosopher import Philosopher

def main(num_philosophers, meals_count):
    # Tworzenie widelców
    forks = [Fork(i) for i in range(1, num_philosophers + 1)]
    philosophers = []

    # Tworzenie filozofów
    for i in range(1, num_philosophers + 1):
        lower_fork = forks[i - 1]
        higher_fork = forks[(i % num_philosophers)] if i % num_philosophers != 0 else forks[0]
        
        # Ustalanie niższego i wyższego widelca
        if lower_fork.id > higher_fork.id:
            lower_fork, higher_fork = higher_fork, lower_fork
        
        philosopher = Philosopher(i, lower_fork, higher_fork, meals_count)
        philosophers.append(philosopher)

    # Startowanie wątków filozofów
    for philosopher in philosophers:
        philosopher.start()

    # Oczekiwanie na zakończenie wątków
    for philosopher in philosophers:
        philosopher.join()


if __name__ == "__main__":
    num_philosophers = 7
    meals_count = 3
    main(num_philosophers, meals_count)
